lib/internal/modules/esm/initialize_import_meta.js

```diff
@@ -14,7 +21,7 @@ function createImportMetaResolve(defaultParentUrl, loader) {
     let url;
 
     try {
-      ({ url } = loader.resolve(specifier, parentUrl));
+      ({ url } = loader.resolve(specifier, parentUrl, importAssertions));
```

Comment:
Forgive my meek understanding – `import.meta.resolve` needs to be sync and so this method has to be sync. I'm piping this knowledge through import assertions with a special symbol `kResolveSync` in order to indicate to upstream that the resolution has to occur synchronously.

If this assertion cannot be met then `import.meta.resolve` cannot be used – as is the case of trying to invoke it inside a loader.

Reply:
Import assertions is the new part of the `import` syntax, like `import data from './data.json' assert { type: 'json' }`. It’s not a place to slip in extra metadata for Node’s internal use.

I assume you’re doing this to allow loaders to somehow know when they’re allowed to call `import.meta.resolve`? I would remove this addition for now and attempt to tackle that in its own PR. We can leave the current restriction of “no `import.meta.resolve` in loaders” in place for now.

Reply:
> I assume you’re doing this to allow loaders to somehow know when they’re allowed to call import.meta.resolve?

Actually no – this is to know whether the _act_ of invoking `resolve` should be synchronous or not. This is a bit of a weird case because `import.meta.resolve` **MUST BE** _synchronous_ and `import()` **MUST BE** asynchronous; both code paths invoke the same `resolve`.

> It’s not a place to slip in extra metadata for Node’s internal use.

This is fair. I just need a way to communicate to the loading system whether or not resolution should be synchronous or asynchronous. The import assertion seemed half-sensible since you can say "this import must be json" you could also kind of say this import must be synchronous.

Reply:
Can we just leave it as always async for now? And leave changes to this area for a future PR.

I understand that this leaves `import.meta.resolve` broken within loaders code, but it’s already broken so that’s fine. I assume nothing changes re `import.meta.resolve` in application code?

Reply:
This codepath is used by _both_ loaders and normal userspace callers so reverting this to async would also break existing normal usage of `import.meta.resolve` that expects a sync result.

Reply:
I resolved this by adding a explicit `resolveSync` method. This I think is both more clear and makes it somewhat more obvious why loader's cannot `import.meta.resolve` since `resolveSync` is not definable for the `Hooks` class without some other kind of sorcery.

Reply:
Okay, but we can’t attach extra metadata to `importAssertions`. You need to find another way to track when the call to the hooks needs to be sync vs async.

Reply:
Correct. It's no longer through `importAssertions`; the method name is what distinguishes it now, not `importAssertions`.

Reply:
Apologies, I hadn’t noticed that you’d changed this.